src/main/java/io/cryostat/diagnostic/DiagnosticsHelper.java

```diff
@@ -0,0 +1,244 @@
+                            } catch (Exception e) {
+                                log.error(e);
+                                return null;
+                            }
+                        })
+                .toList();
+    }
+
+    private ThreadDump convertObject(S3Object object) throws Exception {
+        var req = GetObjectTaggingRequest.builder().bucket(bucket).key(object.key()).build();
+        var tagging = storage.getObjectTagging(req);
+        var list = tagging.tagSet();
+        if (!tagging.hasTagSet() || list.isEmpty()) {
+            throw new Exception("No metadata found");
+        }
+        var decodedList = new ArrayList<Pair<String, String>>();
+        list.forEach(
+                t -> {
+                    var encodedKey = t.key();
+                    var decodedKey =
+                            new String(base64Url.decode(encodedKey), StandardCharsets.UTF_8).trim();
+                    var encodedValue = t.value();
+                    var decodedValue =
+                            new String(base64Url.decode(encodedValue), StandardCharsets.UTF_8)
+                                    .trim();
+                    decodedList.add(Pair.of(decodedKey, decodedValue));
+                });
+        var jvmId =
+                decodedList.stream()
+                        .filter(t -> t.getKey().equals("jvmId"))
+                        .map(Pair::getValue)
+                        .findFirst()
+                        .orElseThrow();
+        var uuid =
+                decodedList.stream()
+                        .filter(t -> t.getKey().equals("uuid"))
+                        .map(Pair::getValue)
+                        .findFirst()
+                        .orElseThrow();
+        // content, jvmid, downloadurl, uuid
+        return new ThreadDump(getThreadDumpContent(uuid), jvmId, downloadUrl(jvmId, uuid), uuid);
+    }
+
+    public String getThreadDumpContent(String uuid) throws IOException {
+        InputStream is = getModel(uuid);
+        return new String(is.readAllBytes(), StandardCharsets.UTF_8);
+    }
+
+    private InputStream getModel(String name) {
+        var req = GetObjectRequest.builder().bucket(bucket).key(name).build();
+        return storage.getObject(req);
+    }
+
+    public ThreadDump addThreadDump(String content, String jvmId) {
+        String uuid = UUID.randomUUID().toString();
+        storage.putObject(
+                PutObjectRequest.builder()
+                        .bucket(bucket)
+                        .key(uuid)
+                        .contentType(MediaType.TEXT_PLAIN)
+                        .tagging(createTagging(jvmId, uuid))
+                        .build(),
+                RequestBody.fromString(content));
+        return new ThreadDump(content, jvmId, downloadUrl(jvmId, uuid), uuid);
+    }
+
+    private Tagging createTagging(String jvmId, String uuid) {
+        var map = Map.of("jvmId", jvmId, "uuid", uuid);
```

Comment:
I don't think the `uuid` needs to be included as tagging/metadata, since it's also used as the object's key already.